### ISSUE TITLE
Uppercase type parameters / generics

### ISSUE THREAD

**Issue by jfrolich** [rescript-lang/syntax#135]
One thing that is still a bit awkward when looking at the ReScript syntax is the type parameters that have a leading `'`. When unfamiliar with the syntax, this was the hardest thing to grok for me because I've never seen anything like that. I would always mentally parse the reason code and was looking for the second `'`. This was when I was slightly interested in Reason and looked at some code samples but didn't look up the syntax.

I think it would be a massive improvement to go for capitalized type parameters. I think there is syntax real estate for that, and it's actually consistent with how typescript does it, so instantly familiar. I think this has also been discussed for Reason 4 but not sure about this (@eduardoRFS). Thoughts?

```ts
type null_undefined<'a> = Js.nullable<'a>
```

vs

```ts
type null_undefined<A> = Js.nullable<A>
```

**Comment by EduardoRFS** [comment]
I'm just not sure that's only about syntax, because type inference will also use `'a`, I'm also not sure how easy is to change this behavior in the compiler. So maybe if that's only on the printer and parsing? So that is still `'a` in the AST?
```ocaml
let f = (a, b) => (a, b)
// will be ('a, 'b) => ('a, 'b)
```

**Comment by johnridesabike** [comment]
FWIW, Ramda’s documentation uses `a` and `b` as where we would use `'a` or `'b` https://ramdajs.com/docs/
```
(a → Boolean) → [a] → Boolean
```

My worry with using `A` and `B` is that they look too much like variants. Newcomers sometimes have trouble with the differences between variant types and variant constructors, and so this is a common newcomer error:
```
type stringOrInt = | Str<string> | Int<int>
type stringArray = array<Str> // <- This doesn’t do what the user thinks it does
```

If people starting seeing working code that looks similar to that, then it may just make it more confusing.

**Comment by alexeygolev** [comment]
This was one of the things that I actually didn't like in the new syntax when it was first "teased". The combination of `<` and `'` is surprisingly hard to parse when reading. I know that we're past the point when the change to `<>` can be discussed (and there are actually reasons why `()` was a better choice in the initial...well...second revision of reason). But I think there might be a need to address the type variables. Capitals might be confusing as we have module names and variant types that are capitalised. 
But, I guess, of all options the least confusing albeit harder to read is `<'a>`
